crates/query-group-macro/src/lib.rs

```diff
@@ -0,0 +1,492 @@
+                        signature.span(),
+                        "Queries must have a return type",
+                    ));
+                };
+
+                if let syn::Type::Path(ref ty_path) = *return_ty {
+                    if matches!(query_kind, QueryKind::Input) {
+                        let field = InputStructField {
+                            name: method_name.to_token_stream(),
+                            ty: ty_path.path.to_token_stream(),
+                        };
+
+                        input_struct_fields.push(field);
+                    }
+                }
+
+                match (query_kind, invoke) {
+                    // input
+                    (QueryKind::Input, None) => {
+                        let query = InputQuery {
+                            signature: method.sig.clone(),
+                            create_data_ident: create_data_ident.clone(),
+                        };
+                        let value = Queries::InputQuery(query);
+                        trait_methods.push(value);
+
+                        let setter = InputSetter {
+                            signature: method.sig.clone(),
+                            return_type: *return_ty.clone(),
+                            create_data_ident: create_data_ident.clone(),
+                        };
+                        setter_trait_methods.push(SetterKind::Plain(setter));
+
+                        let setter = InputSetterWithDurability {
+                            signature: method.sig.clone(),
+                            return_type: *return_ty.clone(),
+                            create_data_ident: create_data_ident.clone(),
+                        };
+                        setter_trait_methods.push(SetterKind::WithDurability(setter));
+                    }
+                    (QueryKind::Interned, None) => {
+                        let interned_struct_path = interned_struct_path.unwrap();
+                        let method = Intern {
+                            signature: signature.clone(),
+                            pat_and_tys: pat_and_tys.clone(),
+                            interned_struct_path: interned_struct_path.clone(),
+                        };
+
+                        trait_methods.push(Queries::Intern(method));
+
+                        let mut method = Lookup {
+                            signature: signature.clone(),
+                            pat_and_tys: pat_and_tys.clone(),
+                            return_ty: *return_ty,
+                            interned_struct_path,
+                        };
+                        method.prepare_signature();
+
+                        lookup_signatures
+                            .push(TraitItem::Fn(make_trait_method(method.signature.clone())));
+                        lookup_methods.push(method);
+                    }
+                    // tracked function without *any* invoke.
+                    (QueryKind::Tracked, None) => {
```

Comment:
Can we fold the two cases for `Tracked`, `TrackedWithSalsaStruct` and `Transparent` like this?

```suggestion
                    (QueryKind::Tracked, invoke) => {
```

Reply:
We can on `QueryKind::Tracked` and `QueryKind::Transparent`. It might be possible to extend this to `TrackedWithSalsaStruct`, but we don't know _at the macro usage site_ whether the parameter is a Salsa struct or not. Could be possible if we attached an annotation to the parameter itself, but I don't think it's a particularly high priority for what is largely a migration aide.